---
title: "Accessible Functions"
output: rmarkdown::html_vignette
author: ["Shixiang Wang", "XiongYi"]
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{Accessible Functions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center"
)
```

Firstly, assume you have already installed `UCSCXenaShiny` package.

```{r setup}
library(UCSCXenaShiny)
```

## Fetch Data

In `XenaShiny`, We provide function to retrieve multi-dimensional data including genomic, epigenomic, transcriptomic, and proteomic data from TCGA and CCLE Pan-Cancer dataset for single identifier (gene/protein).

Check parameters:

```{r}
args(query_value)
```

For TCGA gene expression data, we use Xena dataset with ID `TcgaTargetGtex_rsem_gene_tpm` which includes 19131 samples with tumor tissue samples and normal tissue samples. The expression value unit is `log2(tpm+0.001)`.

Let's check several examples.

### Fetch Gene Expression

```{r}
gene_expr <- query_value("TP53")
```

```{r}
str(gene_expr)
```

### Fetch Transcript Expression

```{r eval=FALSE}
transcript_expr <- query_value("ENST00000000233", data_type = "transcript")
```

### Fetch Gene CNV

```{r eval=FALSE}
gene_cnv <- query_value("TP53", data_type = "cnv")
```

### Fetch Gene Mutation

```{r eval=FALSE}
gene_mut <- query_value("TP53", data_type = "mutation")
```

## Data Visualization

### Visualize Single Gene Expression in PANCAN Dataset

```{r fig.width=12}
vis_toil_TvsN(Gene = "TP53", Mode = "Violinplot", Show.P.value = FALSE, Show.P.label = FALSE)
```

### Compare Gene Expression Level in Single Cancer Type

```{r fig.width=5}
vis_toil_TvsN_cancer(
  Gene = "TP53",
  Mode = "Violinplot",
  Show.P.value = TRUE,
  Show.P.label = TRUE,
  Method = "wilcox.test",
  values = c("#DF2020", "#DDDF21"),
  TCGA.only = FALSE,
  Cancer = "ACC"
)
```

### Compare Gene Expression Level in Different Anatomic Regions

This function needs [gganatogram](https://github.com/jespermaag/gganatogram) package, which is not on CRAN. Please install it before using this function.

```{r fig.width=8, fig.height=5}
if (require("gganatogram")) {
  vis_pancan_anatomy(Gene = "TP53", Gender = c("Female", "Male"), option = "D")
}
```

### Visualize Relationship between Gene Expression and Prognosis in the PANCAN Dataset

```{r fig.width=5, fig.height=6}
vis_unicox_tree(
  Gene = "TP53",
  measure = "OS",
  threshold = 0.5,
  values = c("grey", "#E31A1C", "#377DB8")
)
```
