<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ UCSCXenaShiny</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">UCSCXenaShiny</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/api.html">Examples of Accessible Functions</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/openbiox/UCSCXenaShiny/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h2>
      
      <div class="section-desc"><p>Builtin datasets and functions to load data</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="load_data.html">load_data()</a></code> 
        </dt>
        <dd>Load Dataset Provided by This Package</dd>
      </dl><dl><dt>
          
          <code><a href="TCGA.organ.html">TCGA.organ</a></code> 
        </dt>
        <dd>TCGA: Organ Data</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_clinical.html">tcga_clinical</a></code> 
        </dt>
        <dd>Toil Hub: TCGA Clinical Data</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_surv.html">tcga_surv</a></code> 
        </dt>
        <dd>Toil Hub: TCGA Survival Data</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_genome_instability.html">tcga_genome_instability</a></code> 
        </dt>
        <dd>TCGA: Genome Instability Data</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_gtex.html">tcga_gtex</a></code> 
        </dt>
        <dd>Toil Hub: Merged TCGA GTEx Selected Phenotype</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_purity.html">tcga_purity</a></code> 
        </dt>
        <dd>TCGA: Purity Data</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_subtypes.html">tcga_subtypes</a></code> 
        </dt>
        <dd>TCGA Subtype Data</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_tmb.html">tcga_tmb</a></code> 
        </dt>
        <dd>TCGA: TMB (Tumor Mutation Burden) Data</dd>
      </dl><dl><dt>
          
          <code><a href="tcga_clinical_fine.html">tcga_clinical_fine</a></code> 
        </dt>
        <dd>Toil Hub: Cleaned TCGA Clinical Data for grouping</dd>
      </dl><dl><dt>
          
          <code><a href="ccle_absolute.html">ccle_absolute</a></code> 
        </dt>
        <dd>ABSOLUTE Result of CCLE Database</dd>
      </dl><dl><dt>
          
          <code><a href="ccle_info.html">ccle_info</a></code> 
        </dt>
        <dd>Phenotype Info of CCLE Database</dd>
      </dl><dl><dt>
          
          <code><a href="ccle_info_fine.html">ccle_info_fine</a></code> 
        </dt>
        <dd>Cleaned Phenotype Info of CCLE Database for grouping</dd>
      </dl><dl><dt>
          
          <code><a href="toil_info.html">toil_info</a></code> 
        </dt>
        <dd>Toil Hub: TCGA TARGET GTEX Selected Phenotype</dd>
      </dl><dl><dt>
          
          <code><a href="pcawg_info.html">pcawg_info</a></code> 
        </dt>
        <dd>Phenotype Info of PCAWG Database</dd>
      </dl><dl><dt>
          
          <code><a href="pcawg_purity.html">pcawg_purity</a></code> 
        </dt>
        <dd>Purity Data of PCAWG</dd>
      </dl><dl><dt>
          
          <code><a href="pcawg_info_fine.html">pcawg_info_fine</a></code> 
        </dt>
        <dd>Cleaned Phenotype Info of PCAWG Database for grouping</dd>
      </dl></div><div class="section level2">
      <h2 id="data-query">Data Query<a class="anchor" aria-label="anchor" href="#data-query"></a></h2>
      
      <div class="section-desc"><p>Functions to query data from UCSC Xena data hubs in single molecular level. The functionality of functions below may overlap, we rank them by importance.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="dot-opt_pancan.html">.opt_pancan</a></code> 
        </dt>
        <dd>A default setting for pan-cancer studies</dd>
      </dl><dl><dt>
          
          <code><a href="query_molecule_value.html">query_molecule_value()</a></code> 
        </dt>
        <dd>Get Molecule or Signature Data Values from Dense (Genomic) Matrix Dataset of UCSC Xena Data Hubs</dd>
      </dl><dl><dt>
          
          <code><a href="query_pancan_value.html">query_pancan_value()</a></code> 
        </dt>
        <dd>Query Single Identifier or Signature Value from Pan-cancer Database</dd>
      </dl><dl><dt>
          
          <code><a href="query_general_value.html">query_general_value()</a></code> 
        </dt>
        <dd>download data for shiny general analysis</dd>
      </dl><dl><dt>
          
          <code><a href="query_tcga_group.html">query_tcga_group()</a></code> 
        </dt>
        <dd>Group TPC samples by build-in or custom phenotype and support filtering or merging operations</dd>
      </dl><dl><dt>
          
          <code><a href="query_toil_value_df.html">query_toil_value_df()</a></code> 
        </dt>
        <dd>Obtain ToilHub Info for Single Molecule</dd>
      </dl><dl><dt>
          
          <code><a href="get_pancan_value.html">get_ccle_cn_value()</a></code> <code><a href="get_pancan_value.html">get_ccle_gene_value()</a></code> <code><a href="get_pancan_value.html">get_ccle_protein_value()</a></code> <code><a href="get_pancan_value.html">get_ccle_mutation_status()</a></code> <code><a href="get_pancan_value.html">get_pancan_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_gene_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_transcript_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_protein_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_mutation_status()</a></code> <code><a href="get_pancan_value.html">get_pancan_cn_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_methylation_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_miRNA_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_gene_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_fusion_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_promoter_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_miRNA_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_APOBEC_mutagenesis_value()</a></code> 
        </dt>
        <dd>Fetch Identifier Value from Pan-cancer Dataset</dd>
      </dl><dl><dt>
          
          <code><a href="available_hosts.html">available_hosts()</a></code> 
        </dt>
        <dd>Show Available Hosts</dd>
      </dl></div><div class="section level2">
      <h2 id="analysis-and-visualiation-for-pan-cancer-study">Analysis and Visualiation for Pan-Cancer Study<a class="anchor" aria-label="anchor" href="#analysis-and-visualiation-for-pan-cancer-study"></a></h2>
      
      <div class="section-desc"><p>Functions to analyze and visualize data mainly from TCGA, PCAWG and CCLE</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="tcga_surv_analysis.html">tcga_surv_get()</a></code> <code><a href="tcga_surv_analysis.html">tcga_surv_plot()</a></code> 
        </dt>
        <dd>TCGA Survival Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="analyze_gene_drug_response_asso.html">analyze_gene_drug_response_asso()</a></code> 
        </dt>
        <dd>Analyze Association between Gene (Signature) and Drug Response with CCLE Data</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_drug_response_asso.html">vis_gene_drug_response_asso()</a></code> 
        </dt>
        <dd>Visualize Gene and Drug-Target Association with CCLE Data</dd>
      </dl><dl><dt>
          
          <code><a href="analyze_gene_drug_response_diff.html">analyze_gene_drug_response_diff()</a></code> 
        </dt>
        <dd>Analyze Difference of Drug Response (IC50 Value (uM)) between Gene (Signature) High and Low Expression with CCLE Data</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_drug_response_diff.html">vis_gene_drug_response_diff()</a></code> 
        </dt>
        <dd>Visualize Gene and Drug Response Difference with CCLE Data</dd>
      </dl><dl><dt>
          
          <code><a href="vis_ccle_tpm.html">vis_ccle_tpm()</a></code> 
        </dt>
        <dd>Visualize CCLE Gene Expression</dd>
      </dl><dl><dt>
          
          <code><a href="vis_ccle_gene_cor.html">vis_ccle_gene_cor()</a></code> 
        </dt>
        <dd>Visualize CCLE Gene Expression Correlation</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_TIL_cor.html">vis_gene_TIL_cor()</a></code> 
        </dt>
        <dd>Heatmap for Correlation between Gene and Tumor Immune Infiltration (TIL)</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_immune_cor.html">vis_gene_immune_cor()</a></code> 
        </dt>
        <dd>Heatmap for Correlation between Gene and Immune Signatures</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_cor.html">vis_gene_cor()</a></code> 
        </dt>
        <dd>Visualize Gene-Gene Correlation in TCGA</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_cor_cancer.html">vis_gene_cor_cancer()</a></code> 
        </dt>
        <dd>Visualize Gene-Gene Correlation in a TCGA Cancer Type</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_stemness_cor.html">vis_gene_stemness_cor()</a></code> 
        </dt>
        <dd>Visualize Correlation between Gene and Tumor Stemness</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_tmb_cor.html">vis_gene_tmb_cor()</a></code> 
        </dt>
        <dd>Visualize Correlation between Gene and TMB (Tumor Mutation Burden)</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_msi_cor.html">vis_gene_msi_cor()</a></code> 
        </dt>
        <dd>Visualize Correlation between Gene and MSI (Microsatellite instability)</dd>
      </dl><dl><dt>
          
          <code><a href="vis_gene_pw_cor.html">vis_gene_pw_cor()</a></code> 
        </dt>
        <dd>Visualize Correlation between Gene and Pathway signature Score</dd>
      </dl><dl><dt>
          
          <code><a href="vis_pancan_anatomy.html">vis_pancan_anatomy()</a></code> 
        </dt>
        <dd>Visualize Single Gene Expression in Anatomy Location</dd>
      </dl><dl><dt>
          
          <code><a href="vis_toil_TvsN.html">vis_toil_TvsN()</a></code> 
        </dt>
        <dd>Visualize Pan-cancer TPM (tumor (TCGA) vs Normal (TCGA &amp; GTEx))</dd>
      </dl><dl><dt>
          
          <code><a href="vis_toil_TvsN_cancer.html">vis_toil_TvsN_cancer()</a></code> 
        </dt>
        <dd>Visualize Gene TPM in Single Cancer Type (Tumor (TCGA) vs Normal (TCGA &amp; GTEx))</dd>
      </dl><dl><dt>
          
          <code><a href="vis_toil_Mut.html">vis_toil_Mut()</a></code> 
        </dt>
        <dd>Visualize molecular profile difference between mutation and wild status of queried gene</dd>
      </dl><dl><dt>
          
          <code><a href="vis_toil_Mut_cancer.html">vis_toil_Mut_cancer()</a></code> 
        </dt>
        <dd>Visualize molecular profile difference between mutation and wild status of queried gene in Single Cancer Type</dd>
      </dl><dl><dt>
          
          <code><a href="vis_unicox_tree.html">vis_unicox_tree()</a></code> 
        </dt>
        <dd>Visualize Single Gene Univariable Cox Result from Toil Data Hub</dd>
      </dl><dl><dt>
          
          <code><a href="vis_pcawg_dist.html">vis_pcawg_dist()</a></code> 
        </dt>
        <dd>Visualize molecular profile in PCAWG</dd>
      </dl><dl><dt>
          
          <code><a href="vis_pcawg_gene_cor.html">vis_pcawg_gene_cor()</a></code> 
        </dt>
        <dd>Visualize Gene-Gene Correlation in TCGA</dd>
      </dl><dl><dt>
          
          <code><a href="vis_pcawg_unicox_tree.html">vis_pcawg_unicox_tree()</a></code> 
        </dt>
        <dd>Visualize Single Gene Univariable Cox Result in PCAWG</dd>
      </dl></div><div class="section level2">
      <h2 id="general-analysis-and-visualiation">General Analysis and Visualiation<a class="anchor" aria-label="anchor" href="#general-analysis-and-visualiation"></a></h2>
      
      <div class="section-desc"><p>Functions to analyze and visualize general data, i.e., any valid data from UCSC Xena or users</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="vis_identifier_cor.html">vis_identifier_cor()</a></code> 
        </dt>
        <dd>Visualize Identifier-Identifier Correlation</dd>
      </dl><dl><dt>
          
          <code><a href="vis_identifier_multi_cor.html">vis_identifier_multi_cor()</a></code> 
        </dt>
        <dd>Visualize Correlation for Multiple Identifiers</dd>
      </dl><dl><dt>
          
          <code><a href="vis_identifier_grp_comparison.html">vis_identifier_grp_comparison()</a></code> 
        </dt>
        <dd>Visualize Comparison of an Molecule Identifier between Groups</dd>
      </dl><dl><dt>
          
          <code><a href="vis_identifier_grp_surv.html">vis_identifier_grp_surv()</a></code> 
        </dt>
        <dd>Visualize Identifier Group Survival Difference</dd>
      </dl><dl><dt>
          
          <code><a href="vis_dim_dist.html">vis_dim_dist()</a></code> 
        </dt>
        <dd>Visualize the distribution difference of samples after dimension reduction analysis</dd>
      </dl><dl><dt>
          
          <code><a href="vis_identifier_dim_dist.html">vis_identifier_dim_dist()</a></code> 
        </dt>
        <dd>Visualize the distribution difference of samples after Molecule Identifier dimension reduction analysis</dd>
      </dl><dl><dt>
          
          <code><a href="ezcor.html">ezcor()</a></code> 
        </dt>
        <dd>Run Correlation between Two Variables and Support Group by a Variable</dd>
      </dl><dl><dt>
          
          <code><a href="ezcor_batch.html">ezcor_batch()</a></code> 
        </dt>
        <dd>Run correlation between two variables in a batch mode and support group by a variable</dd>
      </dl><dl><dt>
          
          <code><a href="ezcor_partial_cor.html">ezcor_partial_cor()</a></code> 
        </dt>
        <dd>Run partial correlation</dd>
      </dl></div><div class="section level2">
      <h2 id="eda-report">EDA &amp; Report<a class="anchor" aria-label="anchor" href="#eda-report"></a></h2>
      
      <div class="section-desc"><p>Functions for exploratory data analysis and report.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="mol_quick_analysis.html">mol_quick_analysis()</a></code> 
        </dt>
        <dd>Quick molecule analysis and report generation</dd>
      </dl></div><div class="section level2">
      <h2 id="shiny-app">Shiny App<a class="anchor" aria-label="anchor" href="#shiny-app"></a></h2>
      
      <div class="section-desc"><p>Functions related to Xena Shiny App</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="app_run.html">app_run()</a></code> 
        </dt>
        <dd>Run UCSC Xena Shiny App</dd>
      </dl></div><div class="section level2">
      <h2 id="others">Others<a class="anchor" aria-label="anchor" href="#others"></a></h2>
      
      <div class="section-desc"><p>Other docs</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="UCSCXenaShiny-package.html">UCSCXenaShiny</a></code> <code><a href="UCSCXenaShiny-package.html">UCSCXenaShiny-package</a></code> 
        </dt>
        <dd>Xena Shiny App</dd>
      </dl><dl><dt>
          
          <code><a href="keep_cat_cols.html">keep_cat_cols()</a></code> 
        </dt>
        <dd>Keep Only Columns Used for Sample Selection</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://orcid.org/0000-0001-9855-7357" class="external-link">Shixiang Wang</a>, <a href="https://github.com/lishensuo" class="external-link">Shensuo Li</a>, <a href="https://orcid.org/0000-0002-4370-9824" class="external-link">Yi Xiong</a>, <a href="https://orcid.org/0000-0002-6277-0137" class="external-link">Longfei Zhao</a>, <a href="https://orcid.org/0000-0002-0177-0774" class="external-link">Kai Gu</a>, Yin Li, Fei Zhao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

