<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Visualize the distribution difference of samples after Molecule Identifier dimension reduction analysis — vis_identifier_dim_dist • UCSCXenaShiny</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualize the distribution difference of samples after Molecule Identifier dimension reduction analysis — vis_identifier_dim_dist"><meta name="description" content="NOTE: the dataset must be dense matrix in UCSC Xena data hubs."><meta property="og:description" content="NOTE: the dataset must be dense matrix in UCSC Xena data hubs."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">UCSCXenaShiny</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/api.html">Examples of Accessible Functions</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/openbiox/UCSCXenaShiny/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visualize the distribution difference of samples after Molecule Identifier dimension reduction analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/openbiox/UCSCXenaShiny/blob/master/R/vis_identifier_dim_dist.R" class="external-link"><code>R/vis_identifier_dim_dist.R</code></a></small>
      <div class="d-none name"><code>vis_identifier_dim_dist.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>NOTE: the dataset must be dense matrix in UCSC Xena data hubs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">vis_identifier_dim_dist</span><span class="op">(</span></span>
<span>  dataset <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">grp_df</span>,</span>
<span>  samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return.data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  DR_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PCA"</span>, <span class="st">"UMAP"</span>, <span class="st">"tSNE"</span><span class="op">)</span>,</span>
<span>  add_margin <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"Set1"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dataset">dataset<a class="anchor" aria-label="anchor" href="#arg-dataset"></a></dt>
<dd><p>the dataset to obtain identifiers.</p></dd>


<dt id="arg-ids">ids<a class="anchor" aria-label="anchor" href="#arg-ids"></a></dt>
<dd><p>the molecule identifiers.</p></dd>


<dt id="arg-grp-df">grp_df<a class="anchor" aria-label="anchor" href="#arg-grp-df"></a></dt>
<dd><p>When <code>dataset</code> and <code>id</code> are all not <code>NULL</code>, it should be a <code>data.frame</code> with 2 columns.</p><ul><li><p>The first column refers to sample ID.</p></li>
<li><p>The second column refers to groups indicated in axis X.</p></li>
</ul></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>default is <code>NULL</code>, can be common sample names for two datasets.</p></dd>


<dt id="arg-return-data">return.data<a class="anchor" aria-label="anchor" href="#arg-return-data"></a></dt>
<dd><p>whether to reture the raw meta/matrix data (list) instead of plot</p></dd>


<dt id="arg-dr-method">DR_method<a class="anchor" aria-label="anchor" href="#arg-dr-method"></a></dt>
<dd><p>the dimension reduction method</p></dd>


<dt id="arg-add-margin">add_margin<a class="anchor" aria-label="anchor" href="#arg-add-margin"></a></dt>
<dd><p>the marginal plot (NULL, "density", "boxplot")</p></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>the color setting of RColorBrewer</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a <code>ggplot</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/UCSCXenaTools/" class="external-link">UCSCXenaTools</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> =========================================================================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> UCSCXenaTools version 1.4.8</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Project URL: https://github.com/ropensci/UCSCXenaTools</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Usages: https://cran.r-project.org/web/packages/UCSCXenaTools/vignettes/USCSXenaTools.html</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you use it in published research, please cite:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Wang et al., (2019). The UCSCXenaTools R package: a toolkit for accessing genomics data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   from UCSC Xena platform, from cancer multi-omics to single-cell RNA-seq.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Journal of Open Source Software, 4(40), 1627, https://doi.org/10.21105/joss.01627</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> =========================================================================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                               --Enjoy it--</span>
<span class="r-in"><span><span class="va">expr_dataset</span> <span class="op">&lt;-</span> <span class="st">"TCGA.LUAD.sampleMap/HiSeqV2_percentile"</span></span></span>
<span class="r-in"><span><span class="va">ids</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"KRAS"</span>, <span class="st">"PTEN"</span>, <span class="st">"MDM2"</span>, <span class="st">"CDKN1A"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">cli_dataset</span> <span class="op">&lt;-</span> <span class="st">"TCGA.LUAD.sampleMap/LUAD_clinicalMatrix"</span></span></span>
<span class="r-in"><span><span class="va">cli_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/UCSCXenaTools/reference/XenaGenerate.html" class="external-link">XenaGenerate</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  subset <span class="op">=</span> <span class="va">XenaDatasets</span> <span class="op">==</span> <span class="va">cli_dataset</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://docs.ropensci.org/UCSCXenaTools/reference/XenaQuery.html" class="external-link">XenaQuery</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://docs.ropensci.org/UCSCXenaTools/reference/XenaDownload.html" class="external-link">XenaDownload</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://docs.ropensci.org/UCSCXenaTools/reference/XenaPrepare.html" class="external-link">XenaPrepare</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This will check url status, please be patient.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> All downloaded files will under directory /tmp/RtmpSKqK9x.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The 'trans_slash' option is FALSE, keep same directory structure as Xena.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directories for datasets...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Downloading TCGA.LUAD.sampleMap/LUAD_clinicalMatrix</span>
<span class="r-in"><span><span class="va">grp_df</span> <span class="op">=</span> <span class="va">cli_df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sampleID"</span>, <span class="st">"gender"</span><span class="op">)</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">vis_identifier_dim_dist</span><span class="op">(</span><span class="va">expr_dataset</span>, <span class="va">ids</span>, <span class="va">grp_df</span>, DR_method<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying data of identifier TP53 from dataset TCGA.LUAD.sampleMap/HiSeqV2_percentile for DR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #5</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tried 5 times but failed, please check URL or your internet connection or try it later!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving data to file /tmp/RtmpSKqK9x/UCSCXenaShiny/7bac6e7df6c078cb307b55eb3cc0c512.rds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying data of identifier KRAS from dataset TCGA.LUAD.sampleMap/HiSeqV2_percentile for DR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #5</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tried 5 times but failed, please check URL or your internet connection or try it later!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving data to file /tmp/RtmpSKqK9x/UCSCXenaShiny/7493d9720bf5ffdc743df4da9dc46cf4.rds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying data of identifier PTEN from dataset TCGA.LUAD.sampleMap/HiSeqV2_percentile for DR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #5</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tried 5 times but failed, please check URL or your internet connection or try it later!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving data to file /tmp/RtmpSKqK9x/UCSCXenaShiny/61bd4abf1c20b6812eda21c33f2c8df1.rds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying data of identifier MDM2 from dataset TCGA.LUAD.sampleMap/HiSeqV2_percentile for DR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #5</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tried 5 times but failed, please check URL or your internet connection or try it later!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving data to file /tmp/RtmpSKqK9x/UCSCXenaShiny/b1b235f5007458d1a9eabe92b19a0798.rds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Querying data of identifier CDKN1A from dataset TCGA.LUAD.sampleMap/HiSeqV2_percentile for DR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try querying data #5</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Tried 5 times but failed, please check URL or your internet connection or try it later!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Saving data to file /tmp/RtmpSKqK9x/UCSCXenaShiny/fa7de4460d6e82e06e87c58bd81833ac.rds</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> access data failed, the message is provided below</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Join columns in `x` must be present in the data.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> Problem with `sample`.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://orcid.org/0000-0001-9855-7357" class="external-link">Shixiang Wang</a>, <a href="https://github.com/lishensuo" class="external-link">Shensuo Li</a>, <a href="https://orcid.org/0000-0002-4370-9824" class="external-link">Yi Xiong</a>, <a href="https://orcid.org/0000-0002-6277-0137" class="external-link">Longfei Zhao</a>, <a href="https://orcid.org/0000-0002-0177-0774" class="external-link">Kai Gu</a>, Yin Li, Fei Zhao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

