<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. -->
<meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script>

<title>Function reference â€¢ UCSCXenaShiny</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- Flatly Theme - Light  -->
<link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">

<!-- Darkly Theme - Dark -->
<link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">

<!-- preferably CSS -->
<link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">

  <script src="../darkswitch.js"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>




<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">UCSCXenaShiny</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/api.html">Examples of Accessible Functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
        <li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/openbiox/UCSCXenaShiny/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data" class="hasAnchor"><a href="#section-data" class="anchor"></a>Data</h2>
          <p class="section-desc"><p>Builtin datasets and functions to load data</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="load_data.html">load_data()</a></code> </p>
        </td>
        <td><p>Load Dataset Provided by This Package</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="TCGA.organ.html">TCGA.organ</a></code> </p>
        </td>
        <td><p>TCGA: Organ Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcga_clinical.html">tcga_clinical</a></code> </p>
        </td>
        <td><p>Toil Hub: TCGA Clinical Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcga_surv.html">tcga_surv</a></code> </p>
        </td>
        <td><p>Toil Hub: TCGA Survival Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcga_genome_instability.html">tcga_genome_instability</a></code> </p>
        </td>
        <td><p>TCGA: Genome Instability Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcga_gtex.html">tcga_gtex</a></code> </p>
        </td>
        <td><p>Toil Hub: Merged TCGA GTEx Selected Phenotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcga_purity.html">tcga_purity</a></code> </p>
        </td>
        <td><p>TCGA: Purity Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcga_subtypes.html">tcga_subtypes</a></code> </p>
        </td>
        <td><p>TCGA Subtype Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcga_tmb.html">tcga_tmb</a></code> </p>
        </td>
        <td><p>TCGA: TMB (Tumor Mutation Burden) Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ccle_absolute.html">ccle_absolute</a></code> </p>
        </td>
        <td><p>ABSOLUTE Result of CCLE Database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ccle_info.html">ccle_info</a></code> </p>
        </td>
        <td><p>Phenotype Info of CCLE Database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toil_info.html">toil_info</a></code> </p>
        </td>
        <td><p>Toil Hub: TCGA TARGET GTEX Selected Phenotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pcawg_info.html">pcawg_info</a></code> </p>
        </td>
        <td><p>Phenotype Info of PCAWG Database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pcawg_purity.html">pcawg_purity</a></code> </p>
        </td>
        <td><p>Purity Data of PCAWG</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-query" class="hasAnchor"><a href="#section-data-query" class="anchor"></a>Data Query</h2>
          <p class="section-desc"><p>Functions to query data from UCSC Xena data hubs in single molecular level. The functionality of functions below may overlap, we rank them by importance.</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="query_molecule_value.html">query_molecule_value()</a></code> </p>
        </td>
        <td><p>Get Molecule or Signature Data Values from Dense (Genomic) Matrix Dataset of UCSC Xena Data Hubs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="query_pancan_value.html">query_pancan_value()</a></code> </p>
        </td>
        <td><p>Query Single Identifier or Signature Value from Pan-cancer Database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_pancan_value.html">get_ccle_cn_value()</a></code> <code><a href="get_pancan_value.html">get_ccle_gene_value()</a></code> <code><a href="get_pancan_value.html">get_ccle_protein_value()</a></code> <code><a href="get_pancan_value.html">get_ccle_mutation_status()</a></code> <code><a href="get_pancan_value.html">get_pancan_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_gene_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_transcript_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_protein_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_mutation_status()</a></code> <code><a href="get_pancan_value.html">get_pancan_cn_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_methylation_value()</a></code> <code><a href="get_pancan_value.html">get_pancan_miRNA_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_gene_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_fusion_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_promoter_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_miRNA_value()</a></code> <code><a href="get_pancan_value.html">get_pcawg_APOBEC_mutagenesis_value()</a></code> </p>
        </td>
        <td><p>Fetch Identifier Value from Pan-cancer Dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="query_toil_value_df.html">query_toil_value_df()</a></code> <code><a href="query_toil_value_df.html">query_toil_value_df()</a></code> </p>
        </td>
        <td><p>Obtain ToilHub Info for Single Molecule</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="available_hosts.html">available_hosts()</a></code> </p>
        </td>
        <td><p>Show Available Hosts</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-analysis-and-visualiation-for-pan-cancer-study" class="hasAnchor"><a href="#section-analysis-and-visualiation-for-pan-cancer-study" class="anchor"></a>Analysis and Visualiation for Pan-Cancer Study</h2>
          <p class="section-desc"><p>Functions to analyze and visualize data mainly from TCGA, PCAWG and CCLE</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="tcga_surv_analysis.html">tcga_surv_get()</a></code> <code><a href="tcga_surv_analysis.html">tcga_surv_plot()</a></code> </p>
        </td>
        <td><p>TCGA Survival Analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="analyze_gene_drug_response_asso.html">analyze_gene_drug_response_asso()</a></code> </p>
        </td>
        <td><p>Analyze Association between Gene (Signature) and Drug Response with CCLE Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_drug_response_asso.html">vis_gene_drug_response_asso()</a></code> </p>
        </td>
        <td><p>Visualize Gene and Drug-Target Association with CCLE Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="analyze_gene_drug_response_diff.html">analyze_gene_drug_response_diff()</a></code> </p>
        </td>
        <td><p>Analyze Difference of Drug Response (IC50 Value (uM)) between Gene (Signature) High and Low Expression with CCLE Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_drug_response_diff.html">vis_gene_drug_response_diff()</a></code> </p>
        </td>
        <td><p>Visualize Gene and Drug Response Difference with CCLE Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_ccle_tpm.html">vis_ccle_tpm()</a></code> </p>
        </td>
        <td><p>Visualize CCLE Gene Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_ccle_gene_cor.html">vis_ccle_gene_cor()</a></code> </p>
        </td>
        <td><p>Visualize CCLE Gene Expression Correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_TIL_cor.html">vis_gene_TIL_cor()</a></code> </p>
        </td>
        <td><p>Heatmap for Correlation between Gene and Tumor Immune Infiltration (TIL)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_immune_cor.html">vis_gene_immune_cor()</a></code> </p>
        </td>
        <td><p>Heatmap for Correlation between Gene and Immune Signatures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_cor.html">vis_gene_cor()</a></code> </p>
        </td>
        <td><p>Visualize Gene-Gene Correlation in TCGA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_cor_cancer.html">vis_gene_cor_cancer()</a></code> </p>
        </td>
        <td><p>Visualize Gene-Gene Correlation in a TCGA Cancer Type</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_stemness_cor.html">vis_gene_stemness_cor()</a></code> </p>
        </td>
        <td><p>Visualize Correlation between Gene and Tumor Stemness</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_tmb_cor.html">vis_gene_tmb_cor()</a></code> </p>
        </td>
        <td><p>Visualize Correlation between Gene and TMB (Tumor Mutation Burden)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_gene_msi_cor.html">vis_gene_msi_cor()</a></code> </p>
        </td>
        <td><p>Visualize Correlation between Gene and MSI (Microsatellite instability)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_pancan_anatomy.html">vis_pancan_anatomy()</a></code> </p>
        </td>
        <td><p>Visualize Single Gene Expression in Anatomy Location</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_toil_TvsN.html">vis_toil_TvsN()</a></code> </p>
        </td>
        <td><p>Visualize Pan-cancer TPM (tumor (TCGA) vs Normal (TCGA &amp; GTEx))</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_toil_TvsN_cancer.html">vis_toil_TvsN_cancer()</a></code> </p>
        </td>
        <td><p>Visualize Gene TPM in Single Cancer Type (Tumor (TCGA) vs Normal (TCGA &amp; GTEx))</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_unicox_tree.html">vis_unicox_tree()</a></code> </p>
        </td>
        <td><p>Visualize Single Gene Univariable Cox Result from Toil Data Hub</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_pcawg_dist.html">vis_pcawg_dist()</a></code> </p>
        </td>
        <td><p>Visualize molecular profile in PCAWG</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_pcawg_gene_cor.html">vis_pcawg_gene_cor()</a></code> </p>
        </td>
        <td><p>Visualize Gene-Gene Correlation in TCGA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_pcawg_unicox_tree.html">vis_pcawg_unicox_tree()</a></code> </p>
        </td>
        <td><p>Visualize Single Gene Univariable Cox Result in PCAWG</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-general-analysis-and-visualiation" class="hasAnchor"><a href="#section-general-analysis-and-visualiation" class="anchor"></a>General Analysis and Visualiation</h2>
          <p class="section-desc"><p>Functions to analyze and visualize general data, i.e., any valid data from UCSC Xena or users</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="vis_identifier_cor.html">vis_identifier_cor()</a></code> </p>
        </td>
        <td><p>Visualize Identifier-Identifier Correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_identifier_multi_cor.html">vis_identifier_multi_cor()</a></code> </p>
        </td>
        <td><p>Visualize Correlation for Multiple Identifiers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_identifier_grp_comparison.html">vis_identifier_grp_comparison()</a></code> </p>
        </td>
        <td><p>Visualize Comparison of an Molecule Identifier between Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vis_identifier_grp_surv.html">vis_identifier_grp_surv()</a></code> </p>
        </td>
        <td><p>Visualize Identifier Group Survival Difference</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ezcor.html">ezcor()</a></code> </p>
        </td>
        <td><p>Run Correlation between Two Variables and Support Group by a Variable</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ezcor_batch.html">ezcor_batch()</a></code> </p>
        </td>
        <td><p>Run correlation between two variables in a batch mode and support group by a variable</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ezcor_partial_cor.html">ezcor_partial_cor()</a></code> </p>
        </td>
        <td><p>Run partial correlation</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-shiny-app" class="hasAnchor"><a href="#section-shiny-app" class="anchor"></a>Shiny App</h2>
          <p class="section-desc"><p>Functions related to Xena Shiny App</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="app_run.html">app_run()</a></code> </p>
        </td>
        <td><p>Run UCSC Xena Shiny App</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-others" class="hasAnchor"><a href="#section-others" class="anchor"></a>Others</h2>
          <p class="section-desc"><p>Other docs</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="UCSCXenaShiny.html">UCSCXenaShiny</a></code> </p>
        </td>
        <td><p>Xena Shiny App</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="keep_cat_cols.html">keep_cat_cols()</a></code> </p>
        </td>
        <td><p>Keep Only Columns Used for Sample Selection</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://orcid.org/0000-0001-9855-7357'>Shixiang Wang</a>, <a href='https://orcid.org/0000-0002-4370-9824'>Yi Xiong</a>, <a href='https://orcid.org/0000-0002-6277-0137'>Longfei Zhao</a>, <a href='https://orcid.org/0000-0002-0177-0774'>Kai Gu</a>, Yin Li, Fei Zhao.</p>
</div>

<div class="pkgdown">
  <p>Made with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1, using <a href="https://preferably.amirmasoudabdol.name/?source=footer">preferably</a> template.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


