<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="UCSCXenaShiny">
<title>Examples of Accessible Functions • UCSCXenaShiny</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Examples of Accessible Functions">
<meta property="og:description" content="UCSCXenaShiny">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">UCSCXenaShiny</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/api.html">Examples of Accessible Functions</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/openbiox/UCSCXenaShiny/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Examples of Accessible Functions</h1>
                        <h4 data-toc-skip class="author">Shixiang Wang,
Yi Xiong, Longfei Zhao, Kai Gu, Yin Li, Fei Zhao</h4>
            
            <h4 data-toc-skip class="date">2024-03-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/openbiox/UCSCXenaShiny/blob/HEAD/vignettes/api.Rmd" class="external-link"><code>vignettes/api.Rmd</code></a></small>
      <div class="d-none name"><code>api.Rmd</code></div>
    </div>

    
    
<p>Firstly, assume you have already installed
<strong>UCSCXenaShiny</strong> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/openbiox/UCSCXenaShiny" class="external-link">UCSCXenaShiny</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; =========================================================================================</span></span>
<span><span class="co">#&gt; UCSCXenaShiny version 2.0.0</span></span>
<span><span class="co">#&gt; Project URL: https://github.com/openbiox/UCSCXenaShiny</span></span>
<span><span class="co">#&gt; Usages: https://openbiox.github.io/UCSCXenaShiny/</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; If you use it in published research, please cite:</span></span>
<span><span class="co">#&gt;   Shixiang Wang, Yi Xiong, Longfei Zhao, Kai Gu, Yin Li, Fei Zhao, Jianfeng Li,</span></span>
<span><span class="co">#&gt;   Mingjie Wang, Haitao Wang, Ziyu Tao, Tao Wu, Yichao Zheng, Xuejun Li, Xue-Song Liu,</span></span>
<span><span class="co">#&gt;   UCSCXenaShiny: An R/CRAN Package for Interactive Analysis of UCSC Xena Data, </span></span>
<span><span class="co">#&gt;   Bioinformatics, 2021;, btab561, https://doi.org/10.1093/bioinformatics/btab561.</span></span>
<span><span class="co">#&gt; =========================================================================================</span></span>
<span><span class="co">#&gt;                               --Enjoy it--</span></span></code></pre></div>
<div class="section level2">
<h2 id="fetch-data">Fetch Data<a class="anchor" aria-label="anchor" href="#fetch-data"></a>
</h2>
<p>We provide function to retrieve multi-dimensional data including
genomic, epigenomic, transcriptomic, and proteomic data from TCGA (note,
this actually contains data from TCGA/TARGET/GTEx databases) and CCLE
Pan-Cancer dataset for single identifier (e.g., gene, protein).</p>
<p>Check parameters:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/args.html" class="external-link">args</a></span><span class="op">(</span><span class="va">query_pancan_value</span><span class="op">)</span></span>
<span><span class="co">#&gt; function (molecule, data_type = c("mRNA", "transcript", "protein", </span></span>
<span><span class="co">#&gt;     "mutation", "cnv", "methylation", "miRNA", "fusion", "promoter", </span></span>
<span><span class="co">#&gt;     "APOBEC"), database = c("toil", "ccle", "pcawg"), reset_id = NULL, </span></span>
<span><span class="co">#&gt;     opt_pancan = .opt_pancan) </span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>For TCGA gene expression data, we use Xena dataset with ID
<code>TcgaTargetGtex_rsem_gene_tpm</code> which includes 19131 samples
with tumor tissue samples and normal tissue samples. The expression
value unit is <code>log2(tpm+0.001)</code>.</p>
<p>Let’s check several examples.</p>
<div class="section level3">
<h3 id="fetch-gene-expression">Fetch Gene Expression<a class="anchor" aria-label="anchor" href="#fetch-gene-expression"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/query_pancan_value.html">query_pancan_value</a></span><span class="op">(</span><span class="st">"TP53"</span><span class="op">)</span></span>
<span><span class="co">#&gt; =========================================================================================</span></span>
<span><span class="co">#&gt; UCSCXenaTools version 1.4.8</span></span>
<span><span class="co">#&gt; Project URL: https://github.com/ropensci/UCSCXenaTools</span></span>
<span><span class="co">#&gt; Usages: https://cran.r-project.org/web/packages/UCSCXenaTools/vignettes/USCSXenaTools.html</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; If you use it in published research, please cite:</span></span>
<span><span class="co">#&gt; Wang et al., (2019). The UCSCXenaTools R package: a toolkit for accessing genomics data</span></span>
<span><span class="co">#&gt;   from UCSC Xena platform, from cancer multi-omics to single-cell RNA-seq.</span></span>
<span><span class="co">#&gt;   Journal of Open Source Software, 4(40), 1627, https://doi.org/10.21105/joss.01627</span></span>
<span><span class="co">#&gt; =========================================================================================</span></span>
<span><span class="co">#&gt;                               --Enjoy it--</span></span>
<span><span class="co">#&gt; Try querying data #1</span></span>
<span><span class="co">#&gt; -&gt; Checking if the dataset has probeMap...</span></span>
<span><span class="co">#&gt; -&gt; Done. ProbeMap is found.</span></span>
<span><span class="co">#&gt; Saving data to file /tmp/Rtmp96m644/UCSCXenaShiny/4822f991b17cbdab3831c455be7620b2.rds</span></span>
<span><span class="co">#&gt; More info about dataset please run following commands:</span></span>
<span><span class="co">#&gt;   library(UCSCXenaTools)</span></span>
<span><span class="co">#&gt;   XenaGenerate(subset = XenaDatasets == "TcgaTargetGtex_rsem_gene_tpm") %&gt;% XenaBrowse()</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">gene_expr</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ expression: Named num [1:19131] 4.79 5.89 5.52 4.43 2.38 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:19131] "GTEX-S4Q7-0003-SM-3NM8M" "TCGA-19-1787-01" "TCGA-S9-A7J2-01" "GTEX-QV31-1626-SM-2S1QC" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "label")= chr "gene expression RNAseq"</span></span>
<span><span class="co">#&gt;  $ unit      : chr "log2(tpm+0.001)"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fetch-transcript-expression">Fetch Transcript Expression<a class="anchor" aria-label="anchor" href="#fetch-transcript-expression"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">transcript_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/query_pancan_value.html">query_pancan_value</a></span><span class="op">(</span><span class="st">"ENST00000000233"</span>, data_type <span class="op">=</span> <span class="st">"transcript"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fetch-gene-cnv">Fetch Gene CNV<a class="anchor" aria-label="anchor" href="#fetch-gene-cnv"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_cnv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/query_pancan_value.html">query_pancan_value</a></span><span class="op">(</span><span class="st">"TP53"</span>, data_type <span class="op">=</span> <span class="st">"cnv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fetch-gene-mutation">Fetch Gene Mutation<a class="anchor" aria-label="anchor" href="#fetch-gene-mutation"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_mut</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/query_pancan_value.html">query_pancan_value</a></span><span class="op">(</span><span class="st">"TP53"</span>, data_type <span class="op">=</span> <span class="st">"mutation"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fetch-mirna-mature-strand-expression">Fetch miRNA mature strand expression<a class="anchor" aria-label="anchor" href="#fetch-mirna-mature-strand-expression"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">miRNA_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/query_pancan_value.html">query_pancan_value</a></span><span class="op">(</span><span class="st">"hsa-let-7a-2-3p"</span>, data_type <span class="op">=</span> <span class="st">"miRNA"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-visualization">Data Visualization<a class="anchor" aria-label="anchor" href="#data-visualization"></a>
</h2>
<div class="section level3">
<h3 id="visualize-single-gene-expression-in-pancan-dataset">Visualize Single Gene Expression in PANCAN Dataset<a class="anchor" aria-label="anchor" href="#visualize-single-gene-expression-in-pancan-dataset"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vis_toil_TvsN.html">vis_toil_TvsN</a></span><span class="op">(</span>Gene <span class="op">=</span> <span class="st">"TP53"</span>, Mode <span class="op">=</span> <span class="st">"Violinplot"</span>, Show.P.value <span class="op">=</span> <span class="cn">FALSE</span>, Show.P.label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading cache data /tmp/Rtmp96m644/UCSCXenaShiny/4822f991b17cbdab3831c455be7620b2.rds</span></span>
<span><span class="co">#&gt; More info about dataset please run following commands:</span></span>
<span><span class="co">#&gt;   library(UCSCXenaTools)</span></span>
<span><span class="co">#&gt;   XenaGenerate(subset = XenaDatasets == "TcgaTargetGtex_rsem_gene_tpm") %&gt;% XenaBrowse()</span></span></code></pre></div>
<p><img src="api_files/figure-html/unnamed-chunk-9-1.png" width="1152" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="compare-gene-expression-level-in-single-cancer-type">Compare Gene Expression Level in Single Cancer Type<a class="anchor" aria-label="anchor" href="#compare-gene-expression-level-in-single-cancer-type"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vis_toil_TvsN_cancer.html">vis_toil_TvsN_cancer</a></span><span class="op">(</span></span>
<span>  Gene <span class="op">=</span> <span class="st">"TP53"</span>,</span>
<span>  Mode <span class="op">=</span> <span class="st">"Violinplot"</span>,</span>
<span>  Show.P.value <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Show.P.label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Method <span class="op">=</span> <span class="st">"wilcox.test"</span>,</span>
<span>  values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#DF2020"</span>, <span class="st">"#DDDF21"</span><span class="op">)</span>,</span>
<span>  TCGA.only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  Cancer <span class="op">=</span> <span class="st">"ACC"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading cache data /tmp/Rtmp96m644/UCSCXenaShiny/4822f991b17cbdab3831c455be7620b2.rds</span></span>
<span><span class="co">#&gt; More info about dataset please run following commands:</span></span>
<span><span class="co">#&gt;   library(UCSCXenaTools)</span></span>
<span><span class="co">#&gt;   XenaGenerate(subset = XenaDatasets == "TcgaTargetGtex_rsem_gene_tpm") %&gt;% XenaBrowse()</span></span>
<span><span class="co">#&gt; Counting P value</span></span>
<span><span class="co">#&gt; Counting P value finished</span></span></code></pre></div>
<p><img src="api_files/figure-html/unnamed-chunk-10-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="compare-gene-expression-level-in-different-anatomic-regions">Compare Gene Expression Level in Different Anatomic Regions<a class="anchor" aria-label="anchor" href="#compare-gene-expression-level-in-different-anatomic-regions"></a>
</h3>
<p>This function needs <a href="https://github.com/jespermaag/gganatogram" class="external-link">gganatogram</a>
package, which is not on CRAN. Please install it before using this
function.</p>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"gganatogram"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/vis_pancan_anatomy.html">vis_pancan_anatomy</a></span><span class="op">(</span>Gene <span class="op">=</span> <span class="st">"TP53"</span>, Gender <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Female"</span>, <span class="st">"Male"</span><span class="op">)</span>, option <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
<p><a href="https://gitee.com/ShixiangWang/ImageCollection/raw/master/png/20210322130516.png" class="external-link">Click
to see the output</a></p>
</div>
<div class="section level3">
<h3 id="visualize-relationship-between-gene-expression-and-prognosis-in-the-pancan-dataset">Visualize Relationship between Gene Expression and Prognosis in the
PANCAN Dataset<a class="anchor" aria-label="anchor" href="#visualize-relationship-between-gene-expression-and-prognosis-in-the-pancan-dataset"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vis_unicox_tree.html">vis_unicox_tree</a></span><span class="op">(</span></span>
<span>  Gene <span class="op">=</span> <span class="st">"TP53"</span>,</span>
<span>  measure <span class="op">=</span> <span class="st">"OS"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"#E31A1C"</span>, <span class="st">"#377DB8"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading cache data /tmp/Rtmp96m644/UCSCXenaShiny/4822f991b17cbdab3831c455be7620b2.rds</span></span>
<span><span class="co">#&gt; More info about dataset please run following commands:</span></span>
<span><span class="co">#&gt;   library(UCSCXenaTools)</span></span>
<span><span class="co">#&gt;   XenaGenerate(subset = XenaDatasets == "TcgaTargetGtex_rsem_gene_tpm") %&gt;% XenaBrowse()</span></span>
<span><span class="co">#&gt; Get data value for TP53</span></span></code></pre></div>
<p><img src="api_files/figure-html/unnamed-chunk-11-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="other-functions-and-data">Other Functions and Data<a class="anchor" aria-label="anchor" href="#other-functions-and-data"></a>
</h3>
<p>All exported data and functions are organized at <a href="https://openbiox.github.io/UCSCXenaShiny/reference/index.html">here</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://orcid.org/0000-0001-9855-7357" class="external-link">Shixiang Wang</a>, <a href="https://github.com/lishensuo" class="external-link">Shensuo Li</a>, <a href="https://orcid.org/0000-0002-4370-9824" class="external-link">Yi Xiong</a>, <a href="https://orcid.org/0000-0002-6277-0137" class="external-link">Longfei Zhao</a>, <a href="https://orcid.org/0000-0002-0177-0774" class="external-link">Kai Gu</a>, Yin Li, Fei Zhao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
